core/trino-main/src/main/java/io/trino/operator/window/pattern/SetEvaluator.java

```diff
@@ -0,0 +1,175 @@
+ * `matchedLabels` passed in the next call).
+ * <p>
+ * Also, a full list of positions for a current match is kept, and it can be obtained via the `getAllPositions()`
+ * method. This is for the purpose comparing pattern matching threads in ThreadEquivalence.
+ */
+public class SetEvaluator
+{
+    private static final int DEFAULT_CAPACITY = 10;
+
+    private final Set<Integer> labels;
+    private final boolean running;
+
+    // length of the aggregated prefix of the match
+    private int aggregated;
+
+    // length of the prefix of the match where all applicable positions were identified,
+    // and stored in `allPositions`. During the pattern matching phase it might exceed
+    // the `aggregated` prefix due to `getAllPositions()` calls.
+    private int evaluated;
+
+    // all identified applicable positions for the current match until the `evaluated` position, starting from 0
+    // this list is updated:
+    // - by `resolveNewPositions()`
+    // - by `getAllPositions()` (only during the pattern matching phase)
+    private final IntList allPositions;
+
+    public SetEvaluator(Set<Integer> labels, boolean running)
+    {
+        this.labels = requireNonNull(labels, "labels is null");
+        this.running = running;
+        this.allPositions = new IntList(DEFAULT_CAPACITY);
+    }
+
+    // for copying
+    private SetEvaluator(Set<Integer> labels, boolean running, int aggregated, int evaluated, IntList allPositions)
+    {
+        this.labels = labels;
+        this.running = running;
+        this.aggregated = aggregated;
+        this.evaluated = evaluated;
+        this.allPositions = allPositions;
+    }
+
+    public void reset()
+    {
+        aggregated = 0;
+        evaluated = 0;
+        allPositions.clear();
+    }
+
+    /**
+     * This method is used for resolving positions for aggregation:
+     * - During pattern matching. In this case, the evaluated label has been appended to `matchedLabels`
+     * - When computing row pattern measures after a non-empty match is found.
+     * Search is limited up to the current row in case of RUNNING semantics and to the entire match in case of FINAL semantics.
+     * <p>
+     * TODO If `evaluated` exceeds `aggregated`, we could reuse the pre-evaluated positions. For that,
+     * we need to keep count of all previously returned positions from the `aggregated` prefix.
+     *
+     * @return array of new matching positions since the last call, relative to partition start
+     */
+    public ArrayView resolveNewPositions(int currentRow, ArrayView matchedLabels, int partitionStart, int patternStart)
+    {
+        checkArgument(currentRow >= patternStart && currentRow < patternStart + matchedLabels.length(), "current row is out of bounds of the match");
+        checkState(aggregated <= evaluated && evaluated <= matchedLabels.length(), "SetEvaluator in inconsistent state");
+
+        IntList positions = new IntList(DEFAULT_CAPACITY);
```

Comment:
Isn't this a subset of allPositions? Why not return an ArrayView over a portion of that, instead?

Reply:
`positions` are relative to partition start, and `allPositions` are relative to match start.